### PR TITLE
ARROW-3609: [Gandiva] Convert Gandiva benchmark tests as gbenchmark t…

### PR DESCRIPTION
**Pull request by shyambits2004**
Gandiva benchmark tests are out of gandiva integration testsuite and part of arrow benchmark suite.

```
$ ./release/gandiva-micro_benchmarks 
2018-11-19 15:50:45
Running ./release/gandiva-micro_benchmarks
Run on (8 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
---------------------------------------------------------------------------
Benchmark                                    Time           CPU Iterations
---------------------------------------------------------------------------
TimedTestAdd3/min_time:1.000               863 us        862 us       1650
TimedTestBigNested/min_time:1.000        14881 us      14872 us         97
TimedTestBigNested/min_time:1.000        15055 us      15047 us         91
TimedTestExtractYear/min_time:1.000      10610 us      10604 us        137
TimedTestFilterAdd2/min_time:1.000        1368 us       1367 us       1018
TimedTestFilterLike/min_time:1.000      616936 us     616606 us          2
TimedTestAllocs/min_time:1.000           31298 us      31256 us         44
TimedTestMultiOr/min_time:1.000          25368 us      25364 us         54
TimedTestInExpr/min_time:1.000           13443 us      13441 us        107
```

### CODE REVIEW

File: cpp/src/gandiva/tests/CMakeLists.txt [hunk 1]
```diff
@@ -27,9 +27,16 @@ foreach(lib_type "shared" "static")
...
   add_gandiva_integ_test(hash_test.cc gandiva_${lib_type})
   add_gandiva_integ_test(in_expr_test.cc gandiva_${lib_type})
   add_gandiva_integ_test(null_validity_test.cc gandiva_${lib_type})
 endforeach(lib_type)
+
+set(GANDIVA_BENCHMARK_LINK_LIBRARIES
+  gandiva_static
+)
+
+ADD_ARROW_BENCHMARK(micro_benchmarks)
```
Comment (xhochy) [inline cpp/src/gandiva/tests/CMakeLists.txt]: Add a `gandiva_` prefix to this.

Comment (wesm) [reply cpp/src/gandiva/tests/CMakeLists.txt]: It'd be nice to have a PREFIX argument like ADD_ARROW_TEST

https://issues.apache.org/jira/browse/ARROW-3836

**Comment by wesm** [comment]
It's so nice that GitHub is showing force pushes now

**Comment by wesm** [comment]
Unrelated to this, but we're getting a lot of gflags failures lately on Windows. Could it be related to the recent conda-forge version bump? 

```
"C:\projects\arrow\cpp\build\install.vcxproj" (default target) (1) ->
"C:\projects\arrow\cpp\build\ALL_BUILD.vcxproj" (default target) (3) ->
"C:\projects\arrow\cpp\build\src\arrow\ipc\json-integration-test.vcxproj" (default target) (33) ->
(Link target) -> 
  json-integration-test.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) unsigned int __cdecl google::ParseCommandLineFlags(int *,char * * *,bool)" (__imp_?ParseCommandLineFlags@google@@YAIPEAHPEAPEAPEAD_N@Z) referenced in function main [C:\projects\arrow\cpp\build\src\arrow\ipc\json-integration-test.vcxproj]
  json-integration-test.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) void __cdecl google::ShutDownCommandLineFlags(void)" (__imp_?ShutDownCommandLineFlags@google@@YAXXZ) referenced in function main [C:\projects\arrow\cpp\build\src\arrow\ipc\json-integration-test.vcxproj]
  json-integration-test.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl google::FlagRegisterer::FlagRegisterer<bool>(char const *,char const *,char const *,bool *,bool *)" (__imp_??$?0_N@FlagRegisterer@google@@QEAA@PEBD00PEA_N1@Z) referenced in function "void __cdecl fLB::`dynamic initializer for 'o_integration''(void)" (??__Eo_integration@fLB@@YAXXZ) [C:\projects\arrow\cpp\build\src\arrow\ipc\json-integration-test.vcxproj]
  json-integration-test.obj : error LNK2019: unresolved external symbol "__declspec(dllimport) public: __cdecl google::FlagRegisterer::FlagRegisterer<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > >(char const *,char const *,char const *,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > *,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > *)" (__imp_??$?0V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@FlagRegisterer@google@@QEAA@PEBD00PEAV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@1@Z) referenced in function "void __cdecl fLS::`dynamic initializer for 'o_arrow''(void)" (??__Eo_arrow@fLS@@YAXXZ) [C:\projects\arrow\cpp\build\src\arrow\ipc\json-integration-test.vcxproj]
  C:\projects\arrow\cpp\build\release\Release\json-integration-test.exe : fatal error LNK1120: 4 unresolved externals [C:\projects\arrow\cpp\build\src\arrow\ipc\json-integration-test.vcxproj]
```

**Review by pravindra — APPROVED** [review]
lgtm

**Comment by wesm** [comment]
I just opened https://github.com/apache/arrow/pull/2993. I'd like to add the "gandiva-" prefix before merging this

**Comment by wesm** [comment]
I added the "gandiva" prefix. I tried running these locally, but something is not right

```
$ ./debug/gandiva-micro_benchmarks 
2018-11-19 15:18:32
Running ./debug/gandiva-micro_benchmarks
Run on (8 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
-------------------------------------------------------------------------
Benchmark                                  Time           CPU Iterations
-------------------------------------------------------------------------
TimedTestAdd3/iterations:1           2169927 ns    2169144 ns          1
TimedTestBigNested/iterations:1     15394219 ns   15394107 ns          1
TimedTestBigNested/iterations:1     15697215 ns   15697429 ns          1
TimedTestExtractYear/iterations:1   10789657 ns   10789284 ns          1
TimedTestFilterAdd2/iterations:1     2141252 ns    2140686 ns          1
TimedTestFilterLike/iterations:1   633241800 ns  633253074 ns          1
TimedTestAllocs/iterations:1        52421817 ns   52422539 ns          1
TimedTestMultiOr/iterations:1       25234957 ns   25234687 ns          1
TimedTestInExpr/iterations:1        17494901 ns   17494212 ns          1
```

Running the benchmarks for only a single iteration probably won't yield very high precision results. I found a bug in `TimedEvaluate` that presents when you run the loop multiple times, so fixed that. I also found that `make micro_benchmarks` yielded a broken executable due to a missing dependency on `precompiled`. I also changed to use the `ASSERT_OK` macros so we get higher quality error messages (I had to do this to diagnose the missing dependency on irhelpers.bc). 

This is looking better, have a look

```
$ ./release/gandiva-micro_benchmarks 
2018-11-19 15:50:45
Running ./release/gandiva-micro_benchmarks
Run on (8 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 256K (x4)
  L3 Unified 8192K (x1)
---------------------------------------------------------------------------
Benchmark                                    Time           CPU Iterations
---------------------------------------------------------------------------
TimedTestAdd3/min_time:1.000               863 us        862 us       1650
TimedTestBigNested/min_time:1.000        14881 us      14872 us         97
TimedTestBigNested/min_time:1.000        15055 us      15047 us         91
TimedTestExtractYear/min_time:1.000      10610 us      10604 us        137
TimedTestFilterAdd2/min_time:1.000        1368 us       1367 us       1018
TimedTestFilterLike/min_time:1.000      616936 us     616606 us          2
TimedTestAllocs/min_time:1.000           31298 us      31256 us         44
TimedTestMultiOr/min_time:1.000          25368 us      25364 us         54
TimedTestInExpr/min_time:1.000           13443 us      13441 us        107
```

BTW I'm surprised by https://issues.apache.org/jira/browse/ARROW-3544 every time I switch to release builds. I think compile times can be significantly improved by breaking this into multiple compilation units

**Review by pravindra — APPROVED** [review]
thanks @wesm - your changes lgtm.

**Comment by wesm** [comment]
@shyambits2004 can you let me know your JIRA id so I can assign the issue to you?

**Comment by wesm** [comment]
Found you https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shyamsingh

**Comment by pitrou** [comment]
Thank you @shyambits2004 for doing this!

**Pull request closed without merging**